### PR TITLE
Add golangci-lint for patches

### PR DESCRIPTION
**Pull request by lxiaopei**
Tests done:
lxiaopei@lxiaopei05N6R nsx-operator % make golangci
===> Running golangci (linux) <===
lxiaopei@lxiaopei05N6R nsx-operator % make build
/Users/lxiaopei/go_project/nsx-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."

===> Formatting Go files <===
go vet ./...
GOOS=linux go build -o /Users/lxiaopei/go_project/nsx-operator/bin  -ldflags '' cmd/main.go
lxiaopei@lxiaopei05N6R nsx-operator % make all
/Users/lxiaopei/go_project/nsx-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."

===> Formatting Go files <===
go vet ./...
GOOS=linux go build -o /Users/lxiaopei/go_project/nsx-operator/bin  -ldflags '' cmd/main.go

### CODE REVIEW

File: pkg/nsx/services/securitypolicy/wrap_test.go [hunk 1]
```diff
@@ -163,18 +163,18 @@ func TestSecurityPolicyService_wrapResourceReference(t *testing.T) {
...
-		map[string]data.DataValue{
-			"l4_protocol":   data.NewStringValue("TCP"),
-			"resource_type": data.NewStringValue("L4PortSetServiceEntry"),
-			// adding the following default values to make it easy when compare the existing object from store and the new built object
-			"marked_for_delete": data.NewBooleanValue(false),
-			"overridden":        data.NewBooleanValue(false),
-		},
-	)
-	children = append(children, serviceEntry)
+	// var children []*data.StructValue
```
Comment (timdengyun) [inline pkg/nsx/services/securitypolicy/wrap_test.go:166]: why we comment out this part?

File: .golangci.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      - name: range
+      - name: superfluous-else
+      - name: var-declaration
+      - name: duplicated-imports
+
+linters:
+  disable-all: true
+  enable:
+    - misspell
+    - gofmt
```
Comment (timdengyun) [inline .golangci.yml:26]: seemly, gofumpt is btter than gofmt?

File: Makefile [hunk 3]
```diff
@@ -48,24 +49,47 @@ generate: controller-gen ## Generate code containing DeepCopy, DeepCopyInto, and
...
 .PHONY: build
 build: generate fmt vet ## Build manager binary.
-	go build -o bin/manager cmd/main.go
+	@mkdir -p $(BINDIR)
+	GOOS=linux go build -o $(BINDIR) $(GOFLAGS) -ldflags '$(LDFLAGS)' cmd/main.go
 
 .PHONY: clean
 clean: generate fmt vet ## Build clean binary.
+	@mkdir -p $(BINDIR)
 	go build -o bin/clean cmd_clean/main.go
```
Comment (TaoZou1) [inline Makefile:93]: update Line93 as Line 88?

File: Makefile [hunk 4]
```diff
@@ -84,6 +108,12 @@ docker-push: ## Push docker image with the manager.
 photon:
 	docker build -t github.com/vmware-tanzu/nsx-operator -f build/image/photon/Dockerfile .
 
+.PHONY: cleanall
+cleanall:
+	@rm -rf $(BINDIR)
+	@rm -rf $(GOLANGCI_LINT_BINDIR)
+	@rm -rf $(CURDIR)/.coverage
+
 ##@ Deployment
```
Comment (TaoZou1) [inline Makefile:117]: could we move Line93 
go build -o bin/clean cmd_clean/main.go
to build target?
And rename cleanall to clean

File: pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go [hunk 5]
```diff
@@ -201,7 +201,8 @@ func (r *NSXServiceAccountReconciler) validateRealized(count uint16, ca []byte,
 	// Validate ca at first time
 	// Validate client cert every GCValidationInterval
 	if count == 0 {
-		for _, nsxServiceAccount := range nsxServiceAccountList.Items {
+		for _, account := range nsxServiceAccountList.Items {
+			nsxServiceAccount := account
 			if nsxserviceaccount.IsNSXServiceAccountRealized(&nsxServiceAccount.Status) {
```
Comment (TaoZou1) [inline pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go:206]: not sure what's the update fix? it seems they are the same.

File: pkg/nsx/services/securitypolicy/firewall.go [hunk 6]
```diff
@@ -193,15 +193,20 @@ func (service *SecurityPolicyService) CreateOrUpdateSecurityPolicy(obj *v1alpha1
 		}
 
 		if (finalProjectShares != nil) && len(*finalProjectShares) != 0 {
-			err = service.shareStore.Apply(finalProjectShares)
+			if err = service.shareStore.Apply(finalProjectShares); err != nil {
+				return err
```
Comment (TaoZou1) [inline pkg/nsx/services/securitypolicy/firewall.go:197]: add a log.Error(err, "failed to apply store", "project", finalProjectShares) ?

File: pkg/nsx/services/securitypolicy/firewall.go [hunk 7]
```diff
@@ -385,18 +390,19 @@ func (service *SecurityPolicyService) DeleteSecurityPolicy(obj interface{}, isVp
 			}
 		}
 		if len(nsxProjectGroups) != 0 {
-			err = service.groupStore.Apply(&nsxProjectGroups)
+			if err = service.groupStore.Apply(&nsxProjectGroups); err != nil {
+				return err
```
Comment (TaoZou1) [inline pkg/nsx/services/securitypolicy/firewall.go:394]: add a log.Error()?

File: pkg/nsx/services/securitypolicy/firewall.go (hunk 6 above)
Comment (lxiaopei) [reply pkg/nsx/services/securitypolicy/firewall.go:197]: Added.

File: .golangci.yml (hunk 2 above)
Comment (lxiaopei) [reply .golangci.yml:26]: I see for other open source Golang project, they use gofmt in .golangci.yml. Just keep the same.

File: Makefile (hunk 3 above)
Comment (lxiaopei) [reply Makefile:93]: ok, changed. thanks.

File: Makefile (hunk 4 above)
Comment (lxiaopei) [reply Makefile:117]: how about we build cmd/main.go and cmd_clean/main.go both in "make build" target and change this to "clean" target?

File: pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go (hunk 5 above)
Comment (lxiaopei) [reply pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go:206]: "G601: Implicit memory aliasing in for loop. (gosec)" error detected by golangci-lint tool for the original way.

File: pkg/nsx/services/securitypolicy/firewall.go (hunk 7 above)
Comment (lxiaopei) [reply pkg/nsx/services/securitypolicy/firewall.go:394]: added.

File: pkg/nsx/services/securitypolicy/wrap_test.go (hunk 1 above)
Comment (lxiaopei) [reply pkg/nsx/services/securitypolicy/wrap_test.go:166]: children is not set in tt.args.children, then the "children" variable is not used, which detected by golangci-lint tool.  Could you help check if this is needed? if not, I will delete it.

Comment (timdengyun) [reply pkg/nsx/services/securitypolicy/wrap_test.go:166]: I see. I checked the code, the part can be removed. Thanks

File: cmd_clean/main.go [hunk 8]
```diff
@@ -88,11 +88,17 @@ func main() {
...
+	// 1. failed to validate config
+	// 2. failed to get nsx client
+	// 3. failed to initialize cleanup service
+	// 4. failed to clean up specific resourc
 	var err error
 	var status clean.Status
 	if useExternalHttp {
 		tr := &http.Transport{
 			IdleConnTimeout: 30 * time.Second,
+			// #nosec G402: ignore insecure options
```
Comment (seanpang-vmware) [inline cmd_clean/main.go:101]: what is this comment for?

**Review by seanpang-vmware — APPROVED** [review]
LGTM

File: Makefile [hunk 9]
```diff
@@ -48,25 +49,44 @@ generate: controller-gen ## Generate code containing DeepCopy, DeepCopyInto, and
...
+	@GOOS=linux $(GOLANGCI_LINT_BIN) run -c $(CURDIR)/.golangci.yml --fix
+
+.PHONY: .coverage
+.coverage:
+	mkdir -p $(CURDIR)/.coverage
+
 .PHONY: test
-test: manifests generate fmt vet envtest ## Run tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test -gcflags=all=-l ./... -coverprofile cover.out  ## Prohibit inline optimization when using gomonkey
+test: manifests generate fmt vet envtest .coverage## Run tests .
```
Comment (heypnus) [inline Makefile:80]: `.coverage## Run tests .` can work but I think we'd better change it to `.coverage  ## Run tests.`

File: Makefile [hunk 10]
```diff
@@ -84,6 +104,17 @@ docker-push: ## Push docker image with the manager.
...
+.PHONY: clean
+clean:
+	@rm -rf $(BINDIR)
+	@rm -rf $(GOLANGCI_LINT_BINDIR)
+	@rm -rf $(CURDIR)/.coverage
+
+.PHONY: verify
+verify:
+	@echo "===> Verifying spellings <==="
+	GO=$(GO) $(CURDIR)/hack/verify-spelling.sh
```
Comment (heypnus) [inline Makefile:116]: Where's the `verify-spelling.sh` from? It looks that we didn't put it in the repo?

Comment (lxiaopei) [reply Makefile:116]: will add it。

File: Makefile (hunk 9 above)
Comment (lxiaopei) [reply Makefile:80]: thanks， changed

File: cmd_clean/main.go (hunk 8 above)
Comment (lxiaopei) [reply cmd_clean/main.go:101]: this is for golangci-lint tool skip the check for this,

**Review by seanpang-vmware — APPROVED** [review]

**Pull request merged**